be1-go/hub/base_hub.go

```diff
@@ -0,0 +1,280 @@
+			return
+		}
+
+		if query.Publish.Params.Message.Data.GetAction() == message.DataAction(message.CreateLaoAction) &&
+			query.Publish.Params.Message.Data.GetObject() == message.DataObject(message.LaoObject) {
+			err := h.createLao(*query.Publish)
+			if err != nil {
+				log.Printf("failed to create lao: %v", err)
+				client.SendError(&id, err)
+				return
+			}
+		} else {
+			log.Printf("invalid method: %s", query.GetMethod())
+			client.SendError(&id, &message.Error{
+				Code:        -1,
+				Description: "you may only invoke lao/create on /root",
+			})
+			return
+		}
+
+		status := 0
+		result := message.Result{General: &status}
+		log.Printf("sending result: %+v", result)
+		client.SendResult(id, result)
+		return
+	}
+
+	if channelID[:6] != "/root/" {
+		log.Printf("channel id must begin with /root/")
+		client.SendError(&id, &message.Error{
+			Code:        -2,
+			Description: "channel id must begin with /root/",
+		})
+		return
+	}
+
+	channelID = channelID[6:]
+	h.RLock()
+	channel, ok := h.channelByID[channelID]
+	if !ok {
+		log.Printf("invalid channel id: %s", channelID)
+		client.SendError(&id, &message.Error{
+			Code:        -2,
+			Description: fmt.Sprintf("channel with id %s does not exist", channelID),
+		})
+		return
+	}
+	h.RUnlock()
+
+	method := query.GetMethod()
+	log.Printf("method: %s", method)
+
+	msg := []message.Message{}
+
+	// TODO: use constants
+	switch method {
+	case "subscribe":
+		err = channel.Subscribe(&client, *query.Subscribe)
+	case "unsubscribe":
+		err = channel.Unsubscribe(&client, *query.Unsubscribe)
+	case "publish":
+		err = channel.Publish(*query.Publish)
+	case "message":
+		log.Printf("cannot handle broadcasts right now")
+	case "catchup":
+		msg = channel.Catchup(*query.Catchup)
+		// TODO send catchup response to client
+	}
+
+	if err != nil {
+		log.Printf("failed to process query: %v", err)
+		client.SendError(&id, err)
+		return
+	}
+
+	result := message.Result{}
+
+	if method == "catchup" {
+		result.Catchup = msg
+	} else {
+		general := 0
+		result.General = &general
+	}
+
+	client.SendResult(id, result)
+}
+
+func (h *baseHub) handleMessageFromWitness(incomingMessage *IncomingMessage) {
+	//TODO
+
+}
+
+func (h *baseHub) handleIncomingMessage(incomingMessage *IncomingMessage) {
+	log.Printf("Hub::handleMessageFromClient: %s", incomingMessage.Message)
+
+	switch incomingMessage.Socket.socketType {
+	case ClientSocketType:
+		h.handleMessageFromClient(incomingMessage)
+		return
+	case WitnessSocketType:
+		h.handleMessageFromWitness(incomingMessage)
+		return
+	default:
+		log.Printf("error: invalid socket type")
+		return
+	}
+
+}
+
+func (h *baseHub) createLao(publish message.Publish) error {
+	h.Lock()
+	defer h.Unlock()
+
+	data, ok := publish.Params.Message.Data.(*message.CreateLAOData)
+	if !ok {
+		return &message.Error{
+			Code:        -4,
+			Description: "failed to cast data to CreateLAOData",
+		}
+	}
+
+	encodedID := base64.StdEncoding.EncodeToString(data.ID)
```

Comment:
Are these changing to base64url as well? Is that in a separate PR?

Reply:
I think I can change it in #329  (if this PR is merged first)